From 2e05688f31c76034b75b32ebfb50af3afaeecc3c Mon Sep 17 00:00:00 2001
From: Istio Automation <istio-testing-bot@google.com>
Date: Tue, 9 Nov 2021 08:01:29 -0800
Subject: [release-1.12] Switch port naming for generated Service (#35976)

* Switch port naming for generated Service

Gateway Listener.name is already distinct, so it makes sense to mirror
that exactly. The current scheme may flip-flop depending on unimportant
YAML ordering

* fallback

Co-authored-by: John Howard <howardjohn@google.com>
---
 pilot/pkg/config/kube/gateway/deploymentcontroller.go      | 6 +++++-
 pilot/pkg/config/kube/gateway/deploymentcontroller_test.go | 7 ++++++-
 .../kube/gateway/testdata/deployment/cluster-ip.yaml       | 3 +++
 3 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/pilot/pkg/config/kube/gateway/deploymentcontroller.go b/pilot/pkg/config/kube/gateway/deploymentcontroller.go
index ab12b3c3b9..698bee4b75 100644
--- a/pilot/pkg/config/kube/gateway/deploymentcontroller.go
+++ b/pilot/pkg/config/kube/gateway/deploymentcontroller.go
@@ -291,7 +291,11 @@ func extractServicePorts(gw gateway.Gateway) []corev1.ServicePort {
 			continue
 		}
 		portNums[int32(l.Port)] = struct{}{}
-		name := fmt.Sprintf("%s-%d", strings.ToLower(string(l.Protocol)), i)
+		name := string(l.Name)
+		if name == "" {
+			// Should not happen since name is required, but in case an invalid resource gets in...
+			name = fmt.Sprintf("%s-%d", strings.ToLower(string(l.Protocol)), i)
+		}
 		svcPorts = append(svcPorts, corev1.ServicePort{
 			Name: name,
 			Port: int32(l.Port),
diff --git a/pilot/pkg/config/kube/gateway/deploymentcontroller_test.go b/pilot/pkg/config/kube/gateway/deploymentcontroller_test.go
index 0cfbf4c240..4dc880a830 100644
--- a/pilot/pkg/config/kube/gateway/deploymentcontroller_test.go
+++ b/pilot/pkg/config/kube/gateway/deploymentcontroller_test.go
@@ -68,7 +68,12 @@ func TestConfigureIstioGateway(t *testing.T) {
 					Namespace:   "default",
 					Annotations: map[string]string{"networking.istio.io/service-type": string(corev1.ServiceTypeClusterIP)},
 				},
-				Spec: v1alpha2.GatewaySpec{},
+				Spec: v1alpha2.GatewaySpec{
+					Listeners: []v1alpha2.Listener{{
+						Name: "http",
+						Port: v1alpha2.PortNumber(80),
+					}},
+				},
 			},
 		},
 	}
diff --git a/pilot/pkg/config/kube/gateway/testdata/deployment/cluster-ip.yaml b/pilot/pkg/config/kube/gateway/testdata/deployment/cluster-ip.yaml
index 96dad00cdc..91930770f5 100644
--- a/pilot/pkg/config/kube/gateway/testdata/deployment/cluster-ip.yaml
+++ b/pilot/pkg/config/kube/gateway/testdata/deployment/cluster-ip.yaml
@@ -17,6 +17,9 @@ spec:
   - name: status-port
     port: 15021
     protocol: TCP
+  - name: http
+    port: 80
+    protocol: TCP
   selector:
     istio.io/gateway-name: default
   type: ClusterIP
-- 
2.35.3

